In [1]:
import pandas as pd
from typing import Dict

In [2]:
def analyze_metrics(csv_path: str) -> Dict[str, Dict[str, float]]:
    # Load CSV
    df = pd.read_csv(csv_path)

    # Define metric columns to analyze
    metric_columns = [
        "sequence", "jaccard", "levenshtein", "structure",
        "content", "content2", "f1", "similarity"
    ]

    # 1. Overall averages
    overall_averages = df[metric_columns].mean().round(4).to_dict()

    # 2. Language-level averages
    language_averages = (
        df.groupby("Language")[metric_columns]
        .mean()
        .round(4)
        .to_dict(orient="index")
    )

    return {
        "overall": overall_averages,
        "by_language": language_averages
    }

In [3]:
import pandas as pd
from IPython.display import display

def display_stats(stats: dict, name: str = "Tool"):
    print(f"Evaluation Summary for: {name}\n")

    # Overall average metrics
    overall_df = pd.DataFrame.from_dict(stats["overall"], orient="index", columns=["Overall Average"])
    display(overall_df)

    # Language-level average metrics
    language_df = pd.DataFrame.from_dict(stats["by_language"], orient="index")
    language_df.index.name = "Language"
    display(language_df)


In [4]:
docling_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_docling.csv")
display_stats(docling_stat, "Docling")

Evaluation Summary for: Docling



,Overall Average
sequence,0.2193
jaccard,0.6047
levenshtein,0.3012
structure,0.0652
content,0.8974
content2,0.9100
f1,0.8197
similarity,0.5454


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.1374,0.3519,0.2628,0.0689,0.8370,0.8605,0.4914,0.4300
CS,0.2199,0.5915,0.2849,0.0605,0.9162,0.9217,0.8123,0.5439
DA,0.2406,0.6361,0.3097,0.0699,0.9453,0.9585,0.8685,0.5755
DE,0.2591,0.6500,0.3253,0.0663,0.9478,0.9641,0.8745,0.5839
EL,0.1251,0.3372,0.2085,0.0659,0.4527,0.4817,0.4581,0.3042
EN,0.2387,0.6246,0.3067,0.0643,0.9547,0.9805,0.8729,0.5775
ES,0.2358,0.6425,0.3228,0.0642,0.9583,0.9845,0.8722,0.5829
ET,0.1844,0.6726,0.2933,0.0646,0.9308,0.9247,0.8813,0.5645
FI,0.1761,0.6746,0.3155,0.0622,0.9284,0.9218,0.8783,0.5653


In [5]:
pymupdf_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_unstructures.csv")
display_stats(pymupdf_stat, "unstructured")

Evaluation Summary for: unstructured



,Overall Average
sequence,0.2169
jaccard,0.7030
levenshtein,0.3136
structure,0.0657
content,0.9474
content2,0.9389
f1,0.9160
similarity,0.5859


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.1978,0.6639,0.3158,0.0639,0.9721,0.9663,0.8886,0.5812
CS,0.2017,0.7424,0.2938,0.0628,0.9570,0.9460,0.9402,0.5920
DA,0.2159,0.7021,0.3140,0.0669,0.9728,0.9636,0.9284,0.5948
DE,0.2743,0.7138,0.3303,0.0652,0.9738,0.9656,0.9300,0.6076
EL,0.1668,0.4297,0.3148,0.0660,0.8600,0.8551,0.6577,0.4786
EN,0.2444,0.6969,0.3141,0.0662,0.9817,0.9771,0.9384,0.6027
ES,0.2275,0.7139,0.3300,0.0644,0.9832,0.9802,0.9350,0.6049
ET,0.2013,0.7349,0.2975,0.0662,0.9391,0.9232,0.9361,0.5855
FI,0.2037,0.7379,0.3182,0.0648,0.9386,0.9232,0.9347,0.5887


In [6]:
pymupdf_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_pdfplumber.csv")
display_stats(pymupdf_stat, "pdfplumber")

Evaluation Summary for: pdfplumber



,Overall Average
sequence,0.1222
jaccard,0.4281
levenshtein,0.2723
structure,0.0142
content,0.9265
content2,0.9521
f1,0.6434
similarity,0.4798


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.0763,0.4119,0.2634,0.0132,0.9449,0.9715,0.6263,0.4725
CS,0.1251,0.4445,0.2668,0.0141,0.9157,0.9412,0.6594,0.4810
DA,0.1316,0.4171,0.2769,0.0153,0.9362,0.9611,0.6373,0.4822
DE,0.1243,0.4188,0.2768,0.0144,0.9335,0.9589,0.6370,0.4805
EL,0.0984,0.4140,0.2587,0.0135,0.9306,0.9569,0.6270,0.4713
EN,0.1271,0.3964,0.2773,0.0132,0.9467,0.9729,0.6163,0.4785
ES,0.1193,0.4107,0.2788,0.0132,0.9535,0.9787,0.6257,0.4828
ET,0.1140,0.4541,0.2699,0.0158,0.8931,0.9150,0.6710,0.4761
FI,0.1240,0.4499,0.2707,0.0165,0.8958,0.9182,0.6628,0.4768


In [5]:
pymupdf_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_pymupdf_hybrid.csv")
display_stats(pymupdf_stat, "pymupdf_hybrid")

Evaluation Summary for: pymupdf_hybrid



,Overall Average
sequence,0.1143
jaccard,0.4288
levenshtein,0.2728
structure,0.0146
content,0.9269
content2,0.9588
f1,0.6442
similarity,0.4801


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.0761,0.4123,0.2656,0.0137,0.9406,0.9747,0.6268,0.4728
CS,0.1214,0.4454,0.2666,0.0145,0.9180,0.9496,0.6604,0.4823
DA,0.1185,0.4177,0.2779,0.0156,0.9348,0.9667,0.6382,0.4814
DE,0.1271,0.4198,0.2771,0.0149,0.9316,0.9641,0.6381,0.4818
EL,0.0954,0.4146,0.2602,0.0139,0.9281,0.9613,0.6277,0.4716
EN,0.1191,0.3971,0.2765,0.0136,0.9413,0.9753,0.6173,0.4772
ES,0.1058,0.4113,0.2792,0.0136,0.9478,0.9805,0.6263,0.4806
ET,0.1020,0.4548,0.2708,0.0161,0.9032,0.9304,0.6717,0.4784
FI,0.1039,0.4506,0.2739,0.0168,0.9034,0.9316,0.6636,0.4777


In [4]:
pymupdf_stat = analyze_metrics("/ltstorage/home/4baba/EUR_lex/evaluation_results/test_evaluation_pymupdf_tables.csv")
display_stats(pymupdf_stat, "pymupdf_tables")

Evaluation Summary for: pymupdf_tables



,Overall Average
sequence,0.0479
jaccard,0.3022
levenshtein,0.1426
structure,0.0122
content,0.9268
content2,0.9587
f1,0.6438
similarity,0.4335


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.0455,0.2617,0.0934,0.0115,0.9405,0.9747,0.6264,0.4220
CS,0.0484,0.2741,0.1374,0.0122,0.9179,0.9495,0.6600,0.4285
DA,0.0518,0.3091,0.1515,0.0130,0.9347,0.9666,0.6377,0.4378
DE,0.0564,0.3088,0.1509,0.0122,0.9315,0.9640,0.6376,0.4374
EL,0.0383,0.2586,0.0918,0.0116,0.9280,0.9612,0.6273,0.4167
EN,0.0556,0.3227,0.1527,0.0115,0.9413,0.9752,0.6168,0.4394
ES,0.0412,0.3103,0.1522,0.0114,0.9478,0.9805,0.6259,0.4385
ET,0.0385,0.3278,0.1467,0.0134,0.9031,0.9303,0.6713,0.4330
FI,0.0502,0.3293,0.1471,0.0137,0.9032,0.9314,0.6631,0.4340


In [5]:
pymupdf_stat = analyze_metrics("/ltstorage/home/4baba/EUR_lex/evaluation_results/test_evaluation_pymupdf.csv")
display_stats(pymupdf_stat, "pymupdf")

Evaluation Summary for: pymupdf



,Overall Average
sequence,0.0928
jaccard,0.3239
levenshtein,0.1161
structure,0.0273
content,0.9523
content2,0.9854
f1,0.9635
similarity,0.4945


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.0165,0.0753,0.0511,0.0270,0.9562,0.9914,0.9629,0.4401
CS,0.0784,0.1789,0.1093,0.0283,0.9482,0.9802,0.9665,0.4700
DA,0.0995,0.3800,0.1246,0.0274,0.9540,0.9868,0.9586,0.5044
DE,0.1145,0.3985,0.1289,0.0267,0.9540,0.9883,0.9586,0.5099
EL,0.0166,0.0707,0.0496,0.0266,0.9535,0.9885,0.9659,0.4388
EN,0.1318,0.4999,0.1263,0.0269,0.9582,0.9931,0.9620,0.5283
ES,0.1037,0.4096,0.1309,0.0273,0.9616,0.9952,0.9637,0.5132
ET,0.0745,0.3815,0.1168,0.0282,0.9452,0.9738,0.9636,0.4977
FI,0.0785,0.3892,0.1219,0.0273,0.9437,0.9730,0.9599,0.4990


### Combined Evaluation Summary

| Metric        | Docling | Unstructured | pdfplumber | pymupdf_hybrid | pymupdf_tables | pymupdf | OLMOCR |
|---------------|--------:|-------------:|-----------:| --------------:| --------------:| -------:| ------:|
| sequence      | 0.2193  | 0.2169       | 0.1222     | 0.1143          | 0.0479        | 0.0928 |
| jaccard       | 0.6047  | 0.7030       | 0.4281     | 0.4288          | 0.3022        | 0.3239 |
| levenshtein   | 0.3012  | 0.3136       | 0.2723     | 0.2728          | 0.1426        | 0.1161 |
| structure     | 0.0652  | 0.0657       | 0.0142     | 0.0146          | 0.0122        | 0.0273 |
| content       | 0.8974  | 0.9474       | 0.9265     | 0.9269          | 0.9268        | 0.9523 |
| content2      | 0.9100  | 0.9389       | 0.9521     | 0.9588          | 0.9587        | 0.9854 | 0.9497 |
| f1            | 0.8197  | 0.9160       | 0.6434     | 0.6442          | 0.6438        | 0.9635 |




md tables problem

---------

"veebruaril" appears as "veeruuaril" – a common OCR typo.
"Fondu osa" instead of "Fondi osaks oleva" – again, OCR likely distorted this.
"lirimaa" instead of "Iirimaa"

------
repetition

In [7]:
import pandas as pd

def calculate_averages(csv_path: str):
    """
    Reads the evaluation CSV and calculates:
    - Overall average content2
    - Per-language average content2
    
    Args:
        csv_path (str): Path to the input CSV file.
    """
    # Load the CSV
    df = pd.read_csv(csv_path)

    # Make sure 'content2' is numeric
    df['content2'] = pd.to_numeric(df['content2'], errors='coerce')

    # Calculate overall average
    overall_avg = df['content2'].mean()

    # Calculate per-language averages
    language_avg = df.groupby('Language')['content2'].mean().sort_index()

    print(f"Overall Average content2: {overall_avg:.4f}\n")
    print("Per-Language Averages:")
    print(language_avg.round(4))

    return overall_avg, language_avg

In [8]:
calculate_averages('/ltstorage/home/4baba/EUR_lex/content2.csv')

Overall Average content2: 0.9497

Per-Language Averages:
Language
BG    0.9669
CS    0.9408
DA    0.9717
DE    0.9764
EL    0.8862
EN    0.9904
ES    0.9900
ET    0.9117
FI    0.9156
FR    0.9712
GA    0.9441
HR    0.9491
HU    0.9611
IT    0.9664
LT    0.9197
LV    0.9139
MT    0.8680
NL    0.9874
PL    0.9593
PT    0.9628
RO    0.9754
SK    0.9426
SL    0.9503
SV    0.9705
Name: content2, dtype: float64


(np.float64(0.9496661420672056),
 Language
 BG    0.966922
 CS    0.940820
 DA    0.971684
 DE    0.976435
 EL    0.886155
 EN    0.990422
 ES    0.990045
 ET    0.911652
 FI    0.915638
 FR    0.971169
 GA    0.944084
 HR    0.949094
 HU    0.961118
 IT    0.966372
 LT    0.919663
 LV    0.913877
 MT    0.867951
 NL    0.987388
 PL    0.959272
 PT    0.962793
 RO    0.975380
 SK    0.942596
 SL    0.950317
 SV    0.970492
 Name: content2, dtype: float64)

In [2]:
import pandas as pd
from pathlib import Path

def merge_content2_scores(
    spellcorrected_path: Path,
    pymupdf_path: Path,
    output_path: Path
):
    # Load the spell-corrected content2 scores
    df_spell = pd.read_csv(spellcorrected_path)
    df_spell.rename(columns={"content2": "content2_json_spellCorrected"}, inplace=True)

    # Load the PyMuPDF evaluation results
    df_pymupdf = pd.read_csv(pymupdf_path)
    df_pymupdf.rename(columns={"content2_json": "content2_json", "content2_html": "content2_pymupdf"}, inplace=True)

    # Merge the two datasets on CELEX ID and Language
    merged = pd.merge(df_spell, df_pymupdf, on=["CELEX ID", "Language"], how="inner")

    # Select and reorder desired columns
    merged = merged[[
        "Filepath", "CELEX ID", "Language",
        "content2_json_spellCorrected", "content2_json", "content2_pymupdf"
    ]]

    # Save the final merged result
    merged.to_csv(output_path, index=False)
    print(f"✅ Merged CSV saved to: {output_path}")

In [ ]:
# Usage example
merge_content2_scores(
    spellcorrected_path=Path("/ltstorage/home/4baba/EUR_lex/content2_scores.csv"),
    pymupdf_path=Path("/ltstorage/home/4baba/EUR_lex/evaluation_results/comparison_json_pymupdf.csv"),
    output_path=Path("/ltstorage/home/4baba/EUR_lex/merged_content2_comparison.csv")
)


✅ Merged CSV saved to: /ltstorage/home/4baba/EUR_lex/evaluation_results/merged_content2_comparison.csv
